quark/foreach.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 *--
+
+import os
+import subprocess
+import contextlib
+from argparse import ArgumentParser
+
+from .subproject import Subproject
+from .subproject import SvnSubproject
+
+
+@contextlib.contextmanager
+def change_dir(path):
+    _old_path = os.getcwd()
+
+    try:
+        os.chdir(path)
+        yield
+    finally:
+        os.chdir(_old_path)
+
+
+def run():
+    parser = ArgumentParser(
+        description="""
+        Evaluates an arbitrary shell command in each submodule, skipping all
+        the svn's submodules.
+        The command has access to the variables $name, $sm_path, $displaypath,
+        $sha1 and $toplevel:
+        $name is the name of the submodule;
+        $sm_path is the path of the submodule relative to the superproject;
+        $displaypath is the path of the submodule relative to the root directory;
+        $sha1 is the commit as recorded in the immediate superproject;
+        $toplevel is the absolute path to the top-level of the immediate superproject.
+        """
+    )
+    parser.add_argument(
+        "-q", "--quiet",
+        action="store_true",
+        help="Only print error messages"
+    )
+    parser.add_argument(
+        "command",
+        action="store",
+        nargs="+",
+        help="The command that will be run for every dependency"
+    )
+
+    optlist = parser.parse_args()
+    optlist.source_directory = os.getcwd()
+
+    root, modules = Subproject.create_dependency_tree(
+        optlist.source_directory, update=False
+    )
+
+    for path, module in modules.items():
+        version_control = "svn" if type(module) is SvnSubproject else "git"
```

Comment:
I've just added a `get_version_control` method in Subproject and all the derivates ( GitSubproject and SvnSubproject ).

Reply:
This doesn't change the situation... the gist of the comment above was to avoid, if possible, to have an explicit switch over the project type, and instead add appropriate methods (that would be `virtual` in e.g. C++) in the projects classes to ask for the required information in a type-of-project-independent manner. If that's a complete mess we can avoid this completely and go back to `isinstance`, but it shouldn't really be once we focus what the environment variables to set should be in both cases.

Reply:
Isn't it solved with this commit ? https://github.com/comelz/quark/pull/27/commits/ee383da481b52aec48b86a23ccc30a940ebf8711